File: projects/grind-75/grind-75-python/array/three_sum.py
from typing import List


class Solution:
    @classmethod
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i, num in enumerate(nums):
            if i > 0 and num == nums[i - 1]:
                continue
            l = i + 1
            r = len(nums) - 1
            while l < r:
                total = num + nums[l] + nums[r]
                if total < 0:
                    l += 1
                if total > 0:
                    r -= 1
                if total == 0:
                    res.append((num, nums[l], nums[r]))
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
        return res


"""
Time Complexity - O(nlogn)
Space Complexity - O(n)
"""

File: projects/grind-75/typescript/array/three-sum.ts
export default function threeSum(nums: number[]): number[][] {
  const res: number[][] = [];
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    if (i > 0 && num === nums[i - 1]) continue;
    let l = i + 1;
    let r = nums.length - 1;
    while (l < r) {
      const total = num + nums[l] + nums[r];
      if (total < 0) {
        l++;
      }
      if (total > 0) {
        r--;
      }
      if (total === 0) {
        res.push([num, nums[l], nums[r]]);
        l++;
        r--;
        while (l < r && nums[l] === nums[l - 1]) {
          l++;
        }
        while (l < r && nums[r] === nums[r + 1]) {
          r--;
        }
      }
    }
  }

  return res;
}
